File: packages/frontend/lib/rate-limit.ts
// Law 18: rate-limit middleware is mandatory on auth-bearing endpoints.
// ponytail: in-memory Map, single dev process only -- swap for
// @upstash/ratelimit + @upstash/redis once there's more than one instance.

// Env-configurable so the e2e/ui_verify harness (one fresh login per test,
// ~5 auth requests each, all sharing the same "unknown" key locally) can
// raise the ceiling without touching prod behaviour -- unset/invalid falls
// back to the same 5/60s prod default as before. Read once at module scope.
const DEFAULT_LIMIT = Number(process.env.AUTH_RATE_LIMIT_MAX) || 5;
const DEFAULT_WINDOW_MS = Number(process.env.AUTH_RATE_LIMIT_WINDOW_MS) || 60_000;

/**
 * Sliding-window rate limit check. Records `now` against `key`'s request
 * history in `store`, drops entries older than `windowMs`, and returns
 * whether the request is allowed (true) or should be rejected (false).
 */
export function checkRateLimit(
  store: Map<string, number[]>,
  key: string,
  now: number,
  limit: number = DEFAULT_LIMIT,
  windowMs: number = DEFAULT_WINDOW_MS
): boolean {
  const windowStart = now - windowMs;
  const recent = (store.get(key) ?? []).filter((timestamp) => timestamp > windowStart);

  if (recent.length >= limit) {
    store.set(key, recent);
    return false;
  }

  recent.push(now);
  store.set(key, recent);
  return true;
}

File: packages/backend/src/weave_backend/rate_limit.py
"""Law 18: rate-limit middleware on auth-bearing endpoints.

ponytail: in-memory dict, single dev process -- swap for slowapi/Redis
once there's more than one instance.
"""

from __future__ import annotations


def check_rate_limit(
    store: dict[str, list[float]],
    key: str,
    now: float,
    limit: int = 5,
    window_seconds: float = 60.0,
) -> bool:
    """Sliding-window check: returns True (and records `now`) if `key` has
    made fewer than `limit` requests within the last `window_seconds`.
    """
    window_start = now - window_seconds
    recent = [t for t in store.get(key, []) if t > window_start]
    if len(recent) >= limit:
        store[key] = recent
        return False
    recent.append(now)
    store[key] = recent
    return True
